### PR TITLE
Add thumbnails regeneration and migration scripts

### PR DESCRIPTION
**Pull request by arnaud-morvan**
Relate #7

This contains a huge refactoring.
- Add a storage abstration (local / s3)
- Refactor thumbnails creation and keys manipulation
- Add regenerate and migrate scripts

TODO : Use multithreading in scripts ?

### CODE REVIEW

**Comment by asaunier** [comment]
Could you link the related/fixed issues please?

**Comment by asaunier** [comment]
Thanks! I will make a review shortly.

**Comment by asaunier** [comment]
It seems that you called "thumbnails" all the resized images (BI, MI, SI). Aren't thumbnails only very small images? Isn't it a bit confusing to call "thumbnail" a 800x... image?

**Comment by arnaud-morvan** [comment]
IMHO thumbnail seems good, understandable. It was named like that from the beginning, this is not really in the scope of this PR. Have you another name to propose ?

**Comment by asaunier** [comment]
Could you precise how the migration works and what it does? Perhaps in a small README?

For instance:
* do you only migrate the original images?
* are all existing resized images ("thumbnails") regenerated from the migrated originals? or are they copied from v5?
* where are the v5 images hosted when running the migration script? In the same storage than the final v6 images? Or do we "simply" start by rsync'ing the originals from the v5 storage to the v6 storage and then the thumbnails are regenerated?

Regenerating the "thumbnails" would be really useful to:
* improve their quality
* change (increase!) their sizes

**Comment by asaunier** [comment]
> thumbnail seems good, understandable.

On the other hand when reading your code, I first thought that "thumbnails" were only the small resized images. That's not very intuitive.

It's not a big deal if it's only used in the migration script (because we won't use it much after v6 is live), but a bit more confusing if this naming remains in the application code.

What about "resized_images"?

**Comment by asaunier** [comment]
Shouldn't you also make sure in this PR that SVG images are converted to JPEG instead of PNG?
See https://github.com/c2corg/v6_images/issues/8#issuecomment-250107585

**Comment by arnaud-morvan** [comment]
I do not change anything about svg files. I prefer to make another PR after this is merged.

**Comment by arnaud-morvan** [comment]
I've added scripts description in readme and renamed thumbnail to resized images.

File: README.md [hunk 1]
```diff
@@ -39,13 +39,47 @@ Building and running with Docker
 `make run`
 
 
-Launch migration
------------------------------------
+Launch images migration from V5 to V6
+-------------------------------------
+
+Migration take source images from v5 S3 read-only bucket, which could be
```
Comment (asaunier) [inline README.md]: takeS

File: README.md [hunk 2]
```diff
@@ -39,13 +39,47 @@ Building and running with Docker
...
+V5_ENDPOINT=https://sos.exo.io
+V5_PREFIX: EXO
+EXO_ACCESS_KEY_ID: ...
+EXO_SECRET_KEY: ...
+```
+
+Note that here, ``PREFIX`` point out the keys to use as we can have multiple
+endpoints (AWS, Exoscale) with different keys.
+
+The migration script iterate through v5 images. For each *original* image found:
```
Comment (asaunier) [inline README.md]: iterateS

File: README.md [hunk 3]
```diff
@@ -39,13 +39,47 @@ Building and running with Docker
...
+* If the image already exists on publication bucket, nothing is done (only
+  migrate the new ones).
+* If the image do not exists on the v6 bucket:
```
Comment (asaunier) [inline README.md]: do not exists => does not exist
on the v6 bucket => in the v6 bucket

What's the difference between "v6 bucket" and "publication bucket"?

File: README.md [hunk 4]
```diff
@@ -39,13 +39,47 @@ Building and running with Docker
...
+V5_PREFIX: EXO
+EXO_ACCESS_KEY_ID: ...
+EXO_SECRET_KEY: ...
+```
+
+Note that here, ``PREFIX`` point out the keys to use as we can have multiple
+endpoints (AWS, Exoscale) with different keys.
+
+The migration script iterate through v5 images. For each *original* image found:
+* If the image already exists on publication bucket, nothing is done (only
```
Comment (asaunier) [inline README.md]: on publication bucket => in the publication bucket

File: README.md [hunk 5]
```diff
@@ -39,13 +39,47 @@ Building and running with Docker
...
 
 
-Generate thumbnails after migration
------------------------------------
+Generate *resized* images after migration
+-----------------------------------------
+
+This can be used to change the size or quality of *resized* images.
+
+This script iterate through *published* images. For each *original* image
```
Comment (asaunier) [inline README.md]: iterateS

**Comment by asaunier** [comment]
Thanks for the README and the renaming of thumbnails!
A few minor comments about the README else looks good to me!

File: README.md (hunk 3 above)
Comment (arnaud-morvan) [inline README.md]: We 4 storages : v5_storage, temp_storage, incoming_storage and active_storage.

In the migration script, active_storage is called v6 bucket for the distinction with v5 one.

In the resize_images script is is called publication bucket for the distinction with incoming bucket.

**Comment by asaunier** [comment]
Thanks, feel free to merge if no one else has comment.

**Pull request merged**